### PR TITLE
Mesa demos

### PR DESCRIPTION
**Pull request by camako**
First cut snapified mesa demos showcasing opengl, opengles1, opengles2, egl based apps. Will be used as a tool to develop the missing/incomplete interfaces.

### CODE REVIEW

**Comment by didrocks** [comment]
This pull request is now managed with CodeReviewHub. <a href='https://github.com/ubuntu/snappy-playpen/pull/205'>Refresh</a>.<a href='#crh-update-none'></a>

**Comment by dholbach** [comment]
daniel@daydream:~/dev/snappy/snappy-playpen/mesa-demos$ mesa-demos.es2gears 
/snap/mesa-demos/x1/es2gears: line 13: cd: /snap/mesa-demos/x1/egl/opengles2: No such file or directory
/snap/mesa-demos/x1/es2gears: line 14: /home/daniel/dev/snappy/snappy-playpen/mesa-demos/es2gears_x11: No such file or directory
daniel@daydream:~/dev/snappy/snappy-playpen/mesa-demos$

**Comment by dholbach** [comment]
Could they all share the same wrapper file?

**Comment by dholbach** [comment]
Can you maybe explain why --devmode is still required?

**Comment by camako** [comment]
es2 apps should work now.

**Comment by camako** [comment]
'--devmode' is required because seccomp is denying execution due to a missing syscall (sendmsg) for all these apps. I suspect the helper libraries need them (glx, glut, etc). I'm investigating.

Also, the opengles apps shouldn't use the opengl interface. I'll create an opengles interface in snapd. There might also need to be an egl interface that's needed. Still investigating these and devmode will be removed once these concerns are addressed.

**Comment by camako** [comment]
They could possibly share the same wrapper, but I'm not sure how to express that in the yaml (i.e. copying one file to multiple destinations). I'll ask in the snapcraft list.

But better yet, the wrappers can be eliminated once the 'environment' functionality is merged in (see the readme).

**Comment by camako** [comment]
:+1:

**Comment by dholbach** [comment]
Sorry, when I said "explain" I meant "note it down in the mesa-demos/README.md file".

A way to express it in the yaml file could be something like

apps:
  something:
    command: bin/common-wrapper something
  something-else:
    command: bin/common-wrapper something-else

... if that's possible.

**Comment by dholbach** [comment]
es2gears works now! :-)

**Comment by camako** [comment]
Thanks for your suggestion. Now I use one wrapper. I've also added to the readme why strict mode isn't yet turned on.

**Comment by camako** [comment]
oops

**Comment by camako** [comment]
:+1:

File: mesa-demos/README.md [hunk 1]
```diff
@@ -0,0 +1,27 @@
...
+  - Currently the following applications have been enabled in the snap :
+      teapot
+      gears
+      tunnel2
+      es2info
+      es2tri
+      es2gears
+
+Known issues:
+  - Only validated with Intel (i965) GPU drivers, so far.
```
Comment (dholbach) [inline mesa-demos/README.md]: And `i915`.

File: mesa-demos/README.md [hunk 2]
```diff
@@ -0,0 +1,27 @@
...
+  - Only a few of the demos are included in the snap. Currently, each demo requires an env var to be set which is achieved through a wrapper. Having [lp:1583259](https://bugs.launchpad.net/snappy/+bug/1583259) would help when adding the others wholesale.
+  - 'strict' mode is not yet supported as there are some seccomp denials due to 'sendmsg'. Needs more investigation as to whether a rule addressing that needs to be added to the 'opengl' interface or another interface needs to be created for glut, glx, etc that opengl apps typically use.
```
Comment (dholbach) [inline mesa-demos/README.md]: Is there a snapd/snap-confine bug open for this?

**Comment by dholbach** [comment]
Beautiful work, just added two small comments.

File: mesa-demos/README.md (hunk 2 above)
Comment (camako) [inline mesa-demos/README.md]: I'm still not done investigating this. I'll raise a bug when done.

File: mesa-demos/README.md (hunk 1 above)
Comment (camako) [inline mesa-demos/README.md]: Added it to the readme. Thanks for testing.

File: mesa-demos/README.md (hunk 2 above)
Comment (camako) [inline mesa-demos/README.md]: :+1:<a href='#crh-update-none'></a>

**Comment by camako** [comment]
:+1:<a href='#crh-update-none'></a>

File: mesa-demos/README.md (hunk 1 above)
Comment (camako) [inline mesa-demos/README.md]: :+1:<a href='#crh-update-complete'></a>

**Comment by camako** [comment]
My investigations have lead to this bug : https://bugs.launchpad.net/snappy/+bug/1611978

**Comment by camako** [comment]
Updated the readme to reflect the result of the investigation about the seccomp denial (hence keeping this snap in devmode).

**Comment by camako** [comment]
:+1:

File: mesa-demos/README.md [hunk 3]
```diff
@@ -0,0 +1,27 @@
...
+      es2tri
+      es2gears
+
+Known issues:
+  - Only validated with Intel (i965, i915) GPU drivers, so far.
+  - Only a few of the demos are included in the snap. Currently, each demo requires an env var to be set which is achieved through a wrapper. Having [lp:1583259](https://bugs.launchpad.net/snappy/+bug/1583259) would help when adding the others wholesale.
+  - 'strict' mode is not yet supported due to [lp:1611978](https://bugs.launchpad.net/snappy/+bug/1611978)
+
+TODO:
+ - Validate on post-Xenial
```
Comment (zyga) [inline mesa-demos/README.md]: Validate on Debian, Fedora, Arch, etc...

File: mesa-demos/README.md [hunk 4]
```diff
@@ -0,0 +1,27 @@
...
+ - Validate on non-Intel GPUs
```
Comment (zyga) [inline mesa-demos/README.md]: Validate on various versions of Nvidia proprietary GPU driver (different driver, different behavior from snap-confine!)

**Comment by zyga** [comment]
Looks good. I'd love to use this for my GPU support work in snap-confine. I added two comments that I see as additional TODOs. I would love if those could be included in the tree.

File: mesa-demos/README.md (hunk 3 above)
Comment (camako) [inline mesa-demos/README.md]: I've tried to use an lxd container for Debian and Fedora but they failed for different reasons. I've added this task as a TO-DO to the readme and will get to it when I've completed my main task (which is implementing the opengles interface).

File: mesa-demos/README.md (hunk 4 above)
Comment (camako) [inline mesa-demos/README.md]: Currently I don't have an NVidia based machine. I'll ask around if anyone has one and can help me. For now this task is in the TO-DO.

File: mesa-demos/README.md (hunk 3 above)
Comment (camako) [inline mesa-demos/README.md]: :+1:<a href='#crh-update-none'></a>

File: mesa-demos/README.md (hunk 4 above)
Comment (camako) [inline mesa-demos/README.md]: :+1:<a href='#crh-update-complete'></a>

**Comment by albaguirre** [comment]
Validated on ubuntu 16.04 with nvidia-361 (361.42-0ubuntu2) proprietary driver. Most of the demos included worked fine except mesa-demos.es2tri but that seems to be due to a driver idiosyncrasy rather than the snap itself.

**Comment by camako** [comment]
So can we merge this now?

**Comment by camako** [comment]
I think it fails to build due to an infrastructure problem. I see others failing the same way. How do we rekick the build?

**Comment by dholbach** [comment]
I re-kicked the build. It ran "about 6 hours ago": https://travis-ci.org/ubuntu/snappy-playpen/builds/155385333

**Comment by camako** [comment]
@dholbach @zyga, I've updated the PR with the latest changes in yaml, and also added the missing {build|stage}_packages. Could you give it a quick review again? Thanks.

**Comment by dholbach** [comment]
I'll merge this now. From a Playpen perspective this is well-done and working. If we need to test and further validate this, let's do this in a separate thread.

**Comment by zyga** [comment]
Thank you for merging this. I will look at validating this against snap-confine 1.0.41 that is in yakkety and will soon enter xenial.

**Comment by camako** [comment]
Thanks. I'll create separate PRs for the rest of the demos.

**Pull request merged**